// babel-plugin-transform-modules-ui5/dist/helpers/templates.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.buildInheritingConstructor = exports.buildInheritingFunction = exports.buildThisAssignment = exports.buildExtendAssign = exports.buildNamedImportDestructor = exports.buildConstDeclaration = exports.buildDynamicImportHelper = exports.buildDefaultImportDestructor = exports.buildDefaultImportInterop = exports.buildReturn = exports.buildAllExport = exports.buildAllExportHelper = exports.buildReturnExports = exports.buildTempExport = exports.buildDeclareExports = exports.buildDefineGlobal = exports.buildDefine = exports.buildAssign = exports.exportsIdentifier = undefined;
exports.buildNamedExport = buildNamedExport;

var _babelTypes = require('babel-types');

var t = _interopRequireWildcard(_babelTypes);

var _babelTemplate = require('babel-template');

var _babelTemplate2 = _interopRequireDefault(_babelTemplate);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

const exportName = '__exports';

const exportsIdentifier = exports.exportsIdentifier = t.identifier(exportName);

const buildAssign = exports.buildAssign = (0, _babelTemplate2.default)(`
  OBJECT.NAME = VALUE;
`);

const buildDefine = exports.buildDefine = (0, _babelTemplate2.default)(`
  sap.ui.define([SOURCES], function (PARAMS) {
    BODY;
  });
`);

const buildDefineGlobal = exports.buildDefineGlobal = (0, _babelTemplate2.default)(`
  sap.ui.define([SOURCES], function (PARAMS) {
    BODY;
  }, true);
`);

const buildDeclareExports = exports.buildDeclareExports = (0, _babelTemplate2.default)(`
  const ${ exportName } = {
    __esModule: true
  };
`);

const buildTempExport = exports.buildTempExport = (0, _babelTemplate2.default)(`
  const ${ exportName } = VALUE;
`);

const buildReturnExports = exports.buildReturnExports = (0, _babelTemplate2.default)(`
  return ${ exportName };
`);

// export const buildExportsModuleDeclaration = template(`
//   Object.defineProperty(${exportString}, "__esModule", {
//     value: true
//   });
// `)

function buildNamedExport(obj) {
  // console.log(obj);
  return buildAssign({
    OBJECT: exportsIdentifier,
    NAME: obj.key,
    VALUE: obj.value
  });
}

const buildAllExportHelper = exports.buildAllExportHelper = (0, _babelTemplate2.default)(`
  function extendExports(exports, obj) {
    Object.keys(obj).forEach(function (key) {
      if (key === "default" || key === "__esModule") return;
      Object.defineProperty(exports, key, {
        enumerable: true,
        get: function get() {
          return obj[key];
        }
      });
    });
  }
`);

const buildAllExport = exports.buildAllExport = (0, _babelTemplate2.default)(`
  extendExports(${ exportName }, LOCAL);
`);

const buildReturn = exports.buildReturn = (0, _babelTemplate2.default)(`
  return ID;
`);

const buildDefaultImportInterop = exports.buildDefaultImportInterop = (0, _babelTemplate2.default)(`
  function _interopRequireDefault(obj) { return (obj && obj.__esModule && typeof obj.default !== "undefined") ? obj.default : obj; }
`);

const buildDefaultImportDestructor = exports.buildDefaultImportDestructor = (0, _babelTemplate2.default)(`
  const LOCAL = _interopRequireDefault(MODULE);
`);

// TODO inject __extends instead of Object.assign unless useBuiltIns in set
const buildDynamicImportHelper = exports.buildDynamicImportHelper = (0, _babelTemplate2.default)(`
  function __ui5_require_async(path) {
    return new Promise((resolve, reject) => {
      sap.ui.require([path], (module) => {
        if (!module) { 
          return reject('No module returned from ' + path); 
        } else if (module.__esModule) { 
          return resolve(module); 
        } else if (module.default) {
          return reject(new Error(path + " module includes a 'default' property but not __esModule. Cannot use as dynamic import"));
        } else {
          module.default = typeof module === "object" ? Object.assign({}, module) : module;
          module.__esModule = true;
          resolve(module);
        }
      })
    })
  }
`);

const buildConstDeclaration = exports.buildConstDeclaration = (0, _babelTemplate2.default)(`
  const NAME = VALUE;
`);

const buildNamedImportDestructor = exports.buildNamedImportDestructor = (0, _babelTemplate2.default)(`
  const LOCAL = MODULE[IMPORTED];
`);

const buildExtendAssign = exports.buildExtendAssign = (0, _babelTemplate2.default)(`
  const NAME = SUPERNAME.extend(FQN, OBJECT);
`);

// TODO get this one to use buildAssign
const buildThisAssignment = exports.buildThisAssignment = (0, _babelTemplate2.default)(`
  this.NAME = VALUE;
`);

// export const buildDefaultConstructorFunction = template(`
//   function constructor() {
//     SUPER.prototype.constructor.apply(this, arguments);
//   }
// `)

// This is use when there is not already the function, so always propagate arguments.
const buildInheritingFunction = exports.buildInheritingFunction = (0, _babelTemplate2.default)(`
  function NAME() {
    if (typeof SUPER.prototype.NAME === 'function') {
      SUPER.prototype.NAME.apply(this, arguments);
    }
  }
`);

// This is use when there is not already the function, so always propagate arguments.
const buildInheritingConstructor = exports.buildInheritingConstructor = (0, _babelTemplate2.default)(`
  function constructor() {
    SUPER.prototype.constructor.apply(this, arguments);
  }
`);
